Implement Docker discovery (#8629)
prometheus/prometheus: docs/configuration/configuration.md

dockerswarm_sd_configs:
  [ - <dockerswarm_sd_config> ... ]

# List of Docker service discovery configurations.

> docker_sd_configs:

Reviewer: This needs to be also later on, in the alertmanager_config section.

Author: I've added it to the other section as well. It first looked like other SDs are missing from the alertmanager_config section as well, but it turns out that the list is just sorted a bit differently. Is this something that you want fixed (possibly in a separate PR?)

Reviewer: Yes let's have it in a split PR after this one is merged :)

See [this example Prometheus configuration file](/documentation/examples/prometheus-dockerswarm.yml)
for a detailed example of configuring Prometheus for Docker Swarm.

> ### `<docker_sd_config>`

Reviewer: This should be before dockerswarm_sd_config

> Docker SD configurations allow retrieving scrape targets from [Docker Engine](https://docs.docker.com/engine/) hosts without a Swarm configuration.

Reviewer: ```suggestion
Docker SD configurations allow retrieving scrape targets from [Docker Engine](https://docs.docker.com/engine/) hosts.
```

Available meta labels:

* `__meta_docker_container_id`: the id of the container
* `__meta_docker_container_name`: the name of the container
* `__meta_docker_container_network_mode`: the network mode of the container
* `__meta_docker_container_label_<labelname>`: each label of the container
* `__meta_docker_network_id`: the ID of the network
* `__meta_docker_network_name`: the name of the network
* `__meta_docker_network_ingress`: whether the network is ingress
* `__meta_docker_network_internal`: whether the network is internal
* `__meta_docker_network_label_<labelname>`: each label of the network
* `__meta_docker_network_scope`: the scope of the network
* `__meta_docker_network_ip`: the IP of the container in this network
* `__meta_docker_port_private`: the port on the container
* `__meta_docker_port_public`: the external port if a port-mapping exists

See below for the configuration options for Docker discovery:

```yaml
# Address of the Docker daemon.
host: <string>

# Optional proxy URL.
[ proxy_url: <string> ]

# TLS configuration.
tls_config:
  [ <tls_config> ]

> # The port to scrape metrics from, when `role` is nodes, and for discovered

Reviewer: ```suggestion
# The port to scrape metrics from, when `role` is nodes, and for discovered
```
This is not accurate.
Should be `when no ports are exposed for a container`

Author: I noticed that I completely missed the `port` option in my implementation. The value is currently ignored. Should I remove the `port` configuration option or add it as a fallback for containers which have no exposed ports?

# tasks and services that don't have published ports.
[ port: <int> | default = 80 ]

# Optional filters to limit the discovery process to a subset of available
# resources.
# The available filters are listed in the upstream documentation:

> # Services: https://docs.docker.com/engine/api/v1.40/#operation/ServiceList

Reviewer: This is not the correct documentation link.

# Tasks: https://docs.docker.com/engine/api/v1.40/#operation/TaskList
# Nodes: https://docs.docker.com/engine/api/v1.40/#operation/NodeList
[ filters:
  [ - name: <string>
      values: <string>, [...] ]

> # The time after which the droplets are refreshed.

Reviewer: ```suggestion
# The time after which the containers are refreshed.
```

[ refresh_interval: <duration> | default = 60s ]

# Authentication information used to authenticate to the Docker daemon.
# Note that `basic_auth` and `authorization` options are
# mutually exclusive.
# password and password_file are mutually exclusive.

# Optional HTTP basic authentication information.
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Optional the `Authorization` header configuration.
authorization:
  # Sets the authentication type.
  [ type: <string> | default: Bearer ]
  # Sets the credentials. It is mutually exclusive with
  # `credentials_file`.
  [ credentials: <secret> ]
  # Sets the credentials with the credentials read from the configured file.
  # It is mutually exclusive with `credentials`.
  [ credentials_file: <filename> ]

# Configure whether HTTP requests follow HTTP 3xx redirects.
[ follow_redirects: <bool> | default = true ]

```

The [relabeling phase](#relabel_config) is the preferred and more powerful

> way to filter tasks, services or nodes. For users with thousands of tasks it

Reviewer: ```suggestion
way to filter containers. For users with thousands of containers it
```

can be more efficient to use the Docker API directly which has basic support for

> filtering nodes (using `filters`).

Reviewer: ```suggestion
filtering containers (using `filters`).
```
